### PR TITLE
fix: [k8s-subnetrouter-disapear] change k8s subnetrouter from pod to deployment

### PR DESCRIPTION
**Pull request by antigenius0910**
This PR fix https://github.com/tailscale/tailscale/issues/10405

I am following your k8s tutorial here https://github.com/tailscale/tailscale/blob/main/docs/k8s/README.md for k8s subnet router deployment. 

we notice a problem that seems like there is no `replicaSet`, `statefulSet` or `deployment`  being define in `subnet.yaml` https://github.com/tailscale/tailscale/blob/main/docs/k8s/subnet.yaml. 

therefore when k8s subnet-router node fail or anything bad that can harm to the Pod, then that Pod won't be rescheduled. 

below code will ensure even the k8s worker node rotate subnet-router will still alive and got moved to another k8s worker node inside of the k8s cluster
```
spec:
  replicas: 1
  strategy:
    type: Recreate
```

Update #10405

### CODE REVIEW

**Review by DentonGentry — CHANGES_REQUESTED** [review]
It is not at all clear to me that changing it to be a Deployment is automatically the right thing for all sites.

Marking as Requesting Changes to make it clear that this is not to be submitted yet.

**Comment by antigenius0910** [comment]
> It is not at all clear to me that changing it to be a Deployment is automatically the right thing for all sites.
> 
> Marking as Requesting Changes to make it clear that this is not to be submitted yet.

add more detail into description now @DentonGentry

File: docs/k8s/subnet.yaml [hunk 1]
```diff
@@ -1,31 +1,43 @@
 # Copyright (c) Tailscale Inc & AUTHORS
 # SPDX-License-Identifier: BSD-3-Clause
-apiVersion: v1
-kind: Pod
+apiVersion: apps/v1
+kind: Deployment
```
Comment (irbekrm) [inline docs/k8s/subnet.yaml:4]: In the [Tailscale Kubernetes operator]() we create all standalone proxies as `Statefulset`s (this is mostly for future when we will be able to allow multiple replicas, but will likely need to ensure unique network identifiers and other features that come with `Statefulset`s.) I think we want to make this a `Statefulset` too - for consistency at least.

File: docs/k8s/subnet.yaml [hunk 2]
```diff
@@ -1,31 +1,43 @@
...
-        secretKeyRef:
-          name: tailscale-auth
-          key: TS_AUTHKEY
-          optional: true
-    - name: TS_ROUTES
-      value: "{{TS_ROUTES}}"
-    securityContext:
-      runAsUser: 1000
-      runAsGroup: 1000
+  replicas: 1
```
Comment (irbekrm) [inline docs/k8s/subnet.yaml:10]: Suggestion:

```suggestion
replicas: 1 # Only a single replica is supported
```

File: docs/k8s/subnet.yaml [hunk 3]
```diff
@@ -1,31 +1,43 @@
...
+    metadata:
+      labels:
+        app: tailscale
+    spec:
+      serviceAccountName: "{{SA_NAME}}"
+      containers:
+      - name: tailscale
+        imagePullPolicy: Always
+        image: "ghcr.io/tailscale/tailscale:latest"
+        env:
```
Comment (irbekrm) [inline docs/k8s/subnet.yaml:26]: We will additionally want to set `TS_AUTH_ONCE` env var to "prevent" the node logging in every time the pod is restarted, see https://github.com/tailscale/tailscale/blob/v1.56.1/cmd/k8s-operator/deploy/manifests/proxy.yaml#L31-L32

**Review by irbekrm — CHANGES_REQUESTED** [review]
Hi @antigenius0910 thank you for the PR.

We originally made these manifests just for `Pod`s to signal that these aren't production ready deployments and folks should create `Statefulset`s, `Job`s or whatever else suits according to their needs.
However, I see that things (i.e that replica count currently must be `1`) might be better communicated via a full example `Statefulset`, so this make sense.

I've left a couple comments on the PR. Additionally:
- if we make the `Pod` -> `Statefulset` change for subnet routers, we probably want to do the same for the standalone proxy config in proxy.yaml. If you want to do that in this PR that would make sense, otherwise we can do it separately
- could you please rebase and squash this PR into a single commit?

As a side note, this mechanism for deploying subnet routers and proxies, will likely receive less support in future and the [Tailscale Kubernetes operator](https://tailscale.com/kb/1236/kubernetes-operator) is now the main supported mechanism for exposing cluster routes/components to tailnet and vice versa. (I am going to update these docs with a note about this).

**Comment by antigenius0910** [comment]
@irbekrm 

I have another PR created here with helm https://github.com/tailscale/tailscale/pull/10658
let me close this one.

**Pull request closed without merging**